File: legacy/quantum-api/src/main/java/com/quantum/api/config/SecurityConfig.java
package com.quantum.api.config;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.web.SecurityFilterChain;

/** Spring Security 설정 */
@Configuration
@EnableWebSecurity
public class SecurityConfig {

    @Bean
    public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
        return http.csrf(AbstractHttpConfigurer::disable)
                .authorizeHttpRequests(
                        auth ->
                                auth
                                        // 모니터링 및 헬스체크 경로 허용
                                        .requestMatchers("/actuator/**")
                                        .permitAll()
                                        .requestMatchers("/health")
                                        .permitAll()
                                        .requestMatchers("/metrics")
                                        .permitAll()

                                        // Prometheus 메트릭 경로 허용
                                        .requestMatchers("/prometheus")
                                        .permitAll()
                                        .requestMatchers("/actuator/prometheus")
                                        .permitAll()

                                        // API 문서 경로 허용
                                        .requestMatchers("/swagger-ui/**")
                                        .permitAll()
                                        .requestMatchers("/swagger-ui.html")
                                        .permitAll()
                                        .requestMatchers("/v3/api-docs/**")
                                        .permitAll()
                                        .requestMatchers("/api-docs/**")
                                        .permitAll()
                                        .requestMatchers("/swagger-resources/**")
                                        .permitAll()
                                        .requestMatchers("/webjars/**")
                                        .permitAll()

                                        // 정적 리소스 허용
                                        .requestMatchers("/favicon.ico")
                                        .permitAll()
                                        .requestMatchers("/error")
                                        .permitAll()

                                        // 루트 경로 허용 (Swagger 리다이렉트)
                                        .requestMatchers("/")
                                        .permitAll()
                                        .requestMatchers("/api")
                                        .permitAll()

                                        // 개발용 임시 허용 (추후 인증 구현시 제거)
                                        .requestMatchers("/api/v1/**")
                                        .permitAll()

                                        // 나머지는 인증 필요
                                        .anyRequest()
                                        .authenticated())
                .build();
    }
}

File: quantum-web-api/src/main/kotlin/com/quantum/config/SecurityConfig.kt
package com.quantum.config

import com.quantum.user.infrastructure.security.CustomUserDetailsService
import com.quantum.user.infrastructure.security.JwtAuthenticationFilter
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.AuthenticationProvider
import org.springframework.security.authentication.dao.DaoAuthenticationProvider
import org.springframework.security.config.annotation.authentication.configuration.AuthenticationConfiguration
import org.springframework.security.config.annotation.method.configuration.EnableMethodSecurity
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.http.SessionCreationPolicy
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.web.cors.CorsConfiguration
import org.springframework.web.cors.CorsConfigurationSource
import org.springframework.web.cors.UrlBasedCorsConfigurationSource

/**
 * Spring Security 설정
 */
@Configuration
@EnableWebSecurity
@EnableMethodSecurity(prePostEnabled = true)
class SecurityConfig(
    private val userDetailsService: CustomUserDetailsService,
    private val jwtAuthenticationFilter: JwtAuthenticationFilter
) {
    
    @Bean
    fun passwordEncoder(): PasswordEncoder {
        return BCryptPasswordEncoder()
    }
    
    @Bean
    fun authenticationProvider(): AuthenticationProvider {
        val authProvider = DaoAuthenticationProvider()
        authProvider.setUserDetailsService(userDetailsService)
        authProvider.setPasswordEncoder(passwordEncoder())
        return authProvider
    }
    
    @Bean
    fun authenticationManager(config: AuthenticationConfiguration): AuthenticationManager {
        return config.authenticationManager
    }
    
    @Bean
    fun filterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .csrf { it.disable() }
            .cors { it.configurationSource(corsConfigurationSource()) }
            .sessionManagement { it.sessionCreationPolicy(SessionCreationPolicy.STATELESS) }
            .authorizeHttpRequests { authorize ->
                authorize
                    .requestMatchers("/api/v1/auth/login").permitAll()
                    .requestMatchers("/api/v1/auth/register").permitAll()
                    .requestMatchers("/h2-console/**").permitAll()
                    // KIS 계정 관련 엔드포인트 - 프론트엔드에서 직접 접근
                    .requestMatchers("/api/v1/kis-accounts/me/token").permitAll()
                    .requestMatchers("/api/v1/kis-accounts/me/credentials").permitAll()
                    .requestMatchers("/api/v1/kis-accounts/register").permitAll()
                    // Actuator 엔드포인트 - 모니터링용
                    .requestMatchers("/actuator/health").permitAll()
                    .requestMatchers("/actuator/info").permitAll()
                    .requestMatchers("/actuator/metrics").permitAll()
                    .requestMatchers("/actuator/prometheus").permitAll()
                    // Swagger/OpenAPI 문서화
                    .requestMatchers("/swagger-ui/**").permitAll()
                    .requestMatchers("/swagger-ui.html").permitAll()
                    .requestMatchers("/v3/api-docs/**").permitAll()
                    .requestMatchers("/swagger-resources/**").permitAll()
                    .requestMatchers("/webjars/**").permitAll()
                    .anyRequest().authenticated()
            }
            .authenticationProvider(authenticationProvider())
            .addFilterBefore(jwtAuthenticationFilter, UsernamePasswordAuthenticationFilter::class.java)
            .headers { headers ->
                headers.frameOptions().sameOrigin() // H2 Console 사용을 위해
            }
        
        return http.build()
    }
    
    @Bean
    fun corsConfigurationSource(): CorsConfigurationSource {
        val configuration = CorsConfiguration()
        configuration.allowedOriginPatterns = listOf("*")
        configuration.allowedMethods = listOf("*")
        configuration.allowedHeaders = listOf("*")
        configuration.allowCredentials = true
        
        val source = UrlBasedCorsConfigurationSource()
        source.registerCorsConfiguration("/**", configuration)
        return source
    }
}